### PR TITLE
Collaboration Service

### PR DESCRIPTION
**Pull request by gycgabriel**
REST APIs: 
/rooms/join: Join an existing room using a token.
Method: POST

/rooms/save Save text.
Method: POST

/rooms/:roomId: Retrieve the status or metadata of a specific room.
Method: GET

WebSocket Endpoints:
localhost:5001
“/room/join", (room_id: string, user_id: string) 
"/room/update", (text: string) 
"/room/save", (text: string)
"/room/load", ()

Readme: 
https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g11/tree/collaboration-ts-migrate/services/collaboration-service#readme

### CODE REVIEW

File: frontend/src/pages/collab.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import React, { useEffect, useState } from 'react';
+import { io } from 'socket.io-client';
+
+const socket = io('http://localhost:5001/test');
```
Comment (yhtMinceraft1010X) [inline frontend/src/pages/collab.tsx]: I think port 5001 is for user-service while 5003 should be for the collab service

**Review by yhtMinceraft1010X — COMMENTED** [review]
Generally looks clean, although I have question about the choice of port for the service

File: services/collaboration-service/src/app.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+const server: HTTPServer = http.createServer(app);
+const socketIoOptions: any = {
+  cors: {
+    origin: "http://localhost:3000",
+    methods: ["GET", "POST"],
+  },
+};
+const io: SocketIOServer = new SocketIOServer(server, socketIoOptions);
+
+const PORT: number = parseInt(process.env.PORT || "5001")
```
Comment (ong6) [inline services/collaboration-service/src/app.ts]: should be 5003

File: services/collaboration-service/src/index.html
Comment (ong6) [inline services/collaboration-service/src/index.html:1]: can this be deprecated? should be in the FE instead

**Review by ong6 — APPROVED** [review]
LGTM

File: services/collaboration-service/src/index.html
Comment (gycgabriel) [reply services/collaboration-service/src/index.html:1]: Its an example of how to use it, I think FE needs to change it from html to typescript and Nextjs?

File: frontend/src/pages/collab.tsx (hunk 1 above)
Comment (gycgabriel) [reply frontend/src/pages/collab.tsx]: Ah I see thanks

**Comment by gycgabriel** [comment]
Any comments on API design, does it make sense?

**Comment by gycgabriel** [comment]
Any comments on API design, does it make sense?

File: services/collaboration-service/src/app.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+app.use(express.urlencoded({ extended: false }));
+app.use(cookieParser());
+app.use(express.static(path.join(__dirname, 'public')));
+
+/* Routers */
+app.get('/', function(req, res) {
+  res.sendFile(path.join(__dirname, '/index.html'));
+});
+app.use("/room", roomRouter(io));
+app.use("/docs", swaggerUi.serve, swaggerUi.setup(swaggerFile));
```
Comment (ong6) [inline services/collaboration-service/src/app.ts]: routes should start from /api/service-name

**Comment by ong6** [comment]
> Any comments on API design, does it make sense?

I think overall it looks good, but how does a room get created (is it via load?)

**Comment by gycgabriel** [comment]
> > Any comments on API design, does it make sense?
> 
> I think overall it looks good, but how does a room get created (is it via load?)

From the join api

As long as both users (or more) join the same room_id, which can be a unique concatenation of both user_id or some output from matching service

Load and Save are some extra features I thought of, one-slot saves the text and loads the text

File: services/collaboration-service/src/index.html
Comment (gycgabriel) [reply services/collaboration-service/src/index.html:1]: I think we can leave it as a demo or reference for FE to use

**Pull request merged**